cmd/packer-plugin-check/main.go

```diff
@@ -0,0 +1,178 @@
+// packer-plugin-check is a command used by plugins to validate compatibility and basic configuration
+// to work with Packer.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/packer/packer/plugin"
+)
+
+const packerPluginCheck = "packer-plugin-check"
+
+var (
+	hcl2spec = flag.Bool("hcl2spec", false, "flag to indicate that hcl2spec files should be checked.")
+	docs     = flag.Bool("docs", false, "flag to indicate that documentation files should be checked.")
+	load     = flag.String("load", "", "flag to check if plugin can be loaded by Packer.")
+)
+
+// Usage is a replacement usage function for the flags package.
+func Usage() {
+	fmt.Fprintf(os.Stderr, "Usage of "+packerPluginCheck+":\n")
+	fmt.Fprintf(os.Stderr, "\t"+packerPluginCheck+" [flags]\n")
+	fmt.Fprintf(os.Stderr, "Flags:\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	log.SetFlags(0)
+	log.SetPrefix(packerPluginCheck + ": ")
+	flag.Usage = Usage
+	flag.Parse()
+
+	if *hcl2spec == false && *docs == false && len(*load) == 0 {
+		flag.Usage()
+		os.Exit(2)
+	}
+
+	if *hcl2spec {
+		if err := checkHCL2Specs(); err != nil {
+			fmt.Printf(err.Error())
+			os.Exit(2)
+		}
+		fmt.Printf("Plugin succesfully passed hcl2spec check.\n")
+	}
+
+	if *docs {
+		if err := checkDocumentation(); err != nil {
+			fmt.Printf(err.Error())
+			os.Exit(2)
+		}
+		fmt.Printf("Plugin succesfully passed docs check.\n")
+	}
+
+	if len(*load) > 0 {
+		if err := checkPluginName(*load); err != nil {
+			fmt.Printf(err.Error())
+			os.Exit(2)
+		}
+		if err := discoverAndLoad(); err != nil {
+			fmt.Printf(err.Error())
+			os.Exit(2)
+		}
+		fmt.Printf("Plugin succesfully passed compatibility check.\n")
+	}
+}
+
+// checkHCL2Specs looks for the presence of a hcl2spec.go file in the current directory.
+// It is not possible to predict the number of hcl2spec.go files for a given plugin.
+// Because of that, finding one file is enough to validate the knowledge of hcl2spec generation.
+func checkHCL2Specs() error {
```

Comment:
If there is a missing hclspec file or if the file is out of date will the load check error?

Reply:
No! This only checks if there is at least one hcl2spec file. Is hard to know how many hcl2spec files to expect from a plugin and if they are updated. Anyways, in the plugin goreleaser settings, I added the hcl2spec file generation to update/create any forgotten file. But it will work only if the maintainer sets the hcl2 generation annotation right.

This is kind of to make maintainers aware of the need of generating hcl2spec files.  If they are not aware at all, this will fail.

Reply:
The only way to check that, that I can think of right now, is required plugins to have a hcl2 template example inside the repo so we can verify against the plugin. This is actually something interesting to do! 🤔